dev/WindowsAppRuntime_BootstrapDLL/MddBootstrap.cpp

```diff
@@ -64,55 +79,42 @@ STDAPI MddBootstrapInitialize(
-    CreateLifetimeManager(majorMinorVersion, versionTag, minVersion, lifetimeManager, endTheLifetimeManagerEvent, packageFullName);
-
-    const PACKAGE_INFO* frameworkPackageInfo{};
-    auto packageInfoBuffer{ GetFrameworkPackageInfoForPackage(packageFullName.get(), frameworkPackageInfo) };
-
-    // Temporarily add the framework's package directory to PATH so LoadLibrary can find it and any colocated imports
-    wil::unique_dll_directory_cookie dllDirectoryCookie{ AddFrameworkToPath(frameworkPackageInfo->path) };
-
-    auto windowsAppRuntimeDllFilename{ std::wstring(frameworkPackageInfo->path) + L"\\Microsoft.WindowsAppRuntime.dll" };
-    wil::unique_hmodule windowsAppRuntimeDll(LoadLibraryEx(windowsAppRuntimeDllFilename.c_str(), nullptr, LOAD_WITH_ALTERED_SEARCH_PATH));
-    if (!windowsAppRuntimeDll)
+    // Are we already initialized?
+    auto lock{ std::lock_guard(g_initializationLock) };
+    if (g_initializationCount > 0)
     {
-        const auto lastError{ GetLastError() };
-        THROW_WIN32_MSG(lastError, "Error in LoadLibrary: %d (0x%X) loading %ls", lastError, lastError, windowsAppRuntimeDllFilename.c_str());
+        // Verify the request is compatible with our already initialized state
+        VerifyInitializationIsCompatible(majorMinorVersion, versionTag, minVersion);
+    }
+    else
+    {
+        // First to the key! Do the initialization
+        FirstTimeInitialization(majorMinorVersion, versionTag, minVersion);
     }
 
-    const MddPackageDependencyProcessorArchitectures architectureFilter{};
-    const auto lifetimeKind{ MddPackageDependencyLifetimeKind::Process };
-    const MddCreatePackageDependencyOptions createOptions{};
-    wil::unique_process_heap_string packageDependencyId;
-    THROW_IF_FAILED(MddTryCreatePackageDependency(nullptr, frameworkPackageInfo->packageFamilyName, minVersion, architectureFilter, lifetimeKind, nullptr, createOptions, &packageDependencyId));
-    //
-    const MddAddPackageDependencyOptions addOptions{};
-    MDD_PACKAGEDEPENDENCY_CONTEXT packageDependencyContext{};
-    THROW_IF_FAILED(MddAddPackageDependency(packageDependencyId.get(), MDD_PACKAGE_DEPENDENCY_RANK_DEFAULT, addOptions, &packageDependencyContext, nullptr));
-
-    // Remove out temporary path addition
-    RemoveFrameworkFromPath(frameworkPackageInfo->path);
-    dllDirectoryCookie.reset();
-
-    g_lifetimeManager = lifetimeManager.detach();
-    g_endTheLifetimeManagerEvent = std::move(endTheLifetimeManagerEvent);
-    g_windowsAppRuntimeDll = std::move(windowsAppRuntimeDll);
-    g_packageDependencyId = std::move(packageDependencyId);
-    g_packageDependencyContext = packageDependencyContext;
+    // Success!
+    ++g_initializationCount;
     return S_OK;
 }
 CATCH_RETURN();
 
 STDAPI_(void) MddBootstrapShutdown() noexcept
 {
+    auto lock{ std::lock_guard(g_initializationLock) };
+    const auto initializationCount{ g_initializationCount.load() };
+    if (initializationCount > 1)
```

Comment:
what happens when there is a single initialization? we may want to support completely uninitializing so we can potentially initialize with an otherwise incompatible version?